### ISSUE TITLE
Fail to use 2LO authentication for service account

### ISSUE THREAD

**Issue by wdetac** [nodemailer/nodemailer#871]
I follow the page here: http://nodemailer.com/smtp/oauth2/

Here is my code:

    var transport = {
        port: 587,
        host: "smtp-relay.gmail.com",
        secure: false,
        requireTLS: true,
        auth: {
            type: "OAuth2",
            user: config.client_email,
            serviceClient: config.client_id,
            privateKey: config.private_key
        }
    };

    transporter = nodemailer.createTransport(transport);

    transporter.on('token', token => {
        console.log('A new access token was generated');
        console.log('User: %s', token.user);
        console.log('Access Token: %s', token.accessToken);
        console.log('Expires: %s', new Date(token.expires));
    });

    Promise.try(() => {
            return transporter.verify();
        })
        .then(() => {
            console.log("smtp ready");
        })
        .catch((err) => {
            console.error(err);
        });

Error message:

    {
        code: 'EAUTH',
        response: '535-5.7.8 Username and Password not accepted. Learn more at\n535 5.7.8  https://support.google.com/mail/?p=BadCredentials b72-v6sm1263216iti.0 - gsmtp',
        responseCode: 535,
        command: 'AUTH XOAUTH2'
    }

Access token is created successfully since I have seen `A new access token was generated...` in the console.
I have already read many articles and grant all scope for Gmail as far as I know still I failed.
Note: I successfully send email with 3LO authentication and plain login but I want to use service accounts.
I am using G suite and the service account and enable domain-wide delegation in case it matters.

**Comment by erwin16** [comment]
did you solved it ? I have ecaxtly the same issue at the same point ...exactly

**Comment by erwin16** [comment]
thanks ... I tried to follow this post information ...

https://gypsyshen.com/2018/03/01/secure-gmail-access-with-oauth2-and-2lo-in-express/comment-page-1/#comment-287

but It's failing at the same point, upon . transporter.sendMail() with 401 status code...

**Comment by erwin16** [comment]
Succeeded...  it's running fine .. BUT  the from: email  sender email address is ALWAYS overwritten with the email address used for authentication...

when I use these options 👍 

`
    const mailOptions = {
      from: 'SITE CONTACT<' + sender + '>',
      to: KEY.contact_email,
      subject: 'Message',
      text: 'From: ' + sender + '\n\n' + msg,
      html: '<h1>Message</h1><p>From: ' + sender + '</p><p>' + msg + '</p>'
    };`

we get     from: sender .  being the authenticated user 
and   text:  'From: ' + sender   being the sender email address...

Is there anyway to force , change , bypass this Google modification ?

**Comment by wdetac** [comment]
Try this:
Go to Google Admin Console -> Apps -> G Suite -> Settings for Gmail -> Advanced settings -> SMTP relay service
Change Allowed senders: Only addresses in my domains

**Comment by cloakedninjas** [comment]
@erwin16 How did you solve the 401? I'm getting this at the moment too

**Comment by AaronBaldwin** [comment]
+1, would like to know what the trouble was.  What format of private key data does this expect?  Is there something that needs doing in the google api console?

**Comment by raptoria** [comment]
I got it working, these were the steps:

1. Log in to console.- https://console.cloud.google.com/
2. Create a service account under the project.
3. Click on the new service account, go to permissions and add a member. You will use this member's email address when sending the request.
3. Create keys for the service account. - keys -> add key. https://console.cloud.google.com/iam-admin/serviceaccounts
5. Download your key file. You will get something like service-account-name-accountid_.json. It will have all the information you need to get the code below running.
7. Delegate authority to your service account https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority. Add`https://mail.google.com/` as the scope.
9. Write some code like below:
```

const nodemailer = require('nodemailer');
const json = require('./service-account-name-accountid_.json');

const sendEmail = async (email, subject, text) => {
    try {

        const transporter = nodemailer.createTransport({
            host: 'smtp.gmail.com',
            port: 465,
            secure: true,
            auth: {
                type: 'OAuth2',
                user: email,
                serviceClient: json.client_id,
                privateKey: json.private_key
            }
        });

        await transporter.verify();
        
        await transporter.sendMail({
                from: json.service_email,
                to: email,
                subject,
                text
        });
        console.log('success!');
        return {
            status : 200
        }

    } catch (error) {
        console.log(error);
        return {
            status : 500,
            error
        }
    }
}

sendEmail('your_permissioned_service_account_email_address@some_place.com, 'testing 123', 'woohoo!');
```

**Comment by rhanni** [comment]
Thank you so much @raptoria! I've been trying to figure this out for weeks now. The above solution worked for me - but I had to change this part:
>         await transporter.sendMail({
>                 from: json.service_email,
>                 to: email, //you can change this to any other e-mail address and it should work!
>                 subject,
>                 text
>         }); 

`from: json.service_email` to `from: json.client_email`. `service_email` did not exist in the json file downloaded from the Google Console, but `client_email` did. Keys in the file:
`type`, 
`project_id`, 
`private_key_id`,
`private_key`,
`client_email`,
`client_id`,
`auth_uri`,
`token_uri`,
`auth_provider_x509_cert_url`,
`client_x509_cert_url`

**Comment by raptoria** [comment]
good catch! I updated my answer

**Comment by DrNguyen2525** [comment]
> I got it working, these were the steps:
> 
> 1. Log in to console.- https://console.cloud.google.com/
> 2. Create a service account under the project.
> 3. Click on the new service account, go to permissions and add a member. You will use this member's email address when sending the request.
> 4. Create keys for the service account. - keys -> add key. https://console.cloud.google.com/iam-admin/serviceaccounts
> 5. Download your key file. You will get something like service-account-name-accountid.json. It will have all the information you need to get the code below running.
> 6. Delegate authority to your service account https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority. Add`https://mail.google.com/` as the scope.
> ...

I have tried your solution, but the 6th step seems to require a Google Workspace account to perform. I only have a Regular Gmail account, so this is unavailable for me.

**Comment by gigxpad** [comment]
> I got it working, these were the steps:
> 
> 1. Log in to console.- https://console.cloud.google.com/
> 2. Create a service account under the project.
> 3. Click on the new service account, go to permissions and add a member. You will use this member's email address when sending the request.
> 4. Create keys for the service account. - keys -> add key. https://console.cloud.google.com/iam-admin/serviceaccounts
> 5. Download your key file. You will get something like service-account-name-accountid.json. It will have all the information you need to get the code below running.
> 6. Delegate authority to your service account https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority. Add`https://mail.google.com/` as the scope.
> 7. Write some code like below:
> 
> ```js
> const nodemailer = require('nodemailer');
> const json = require('./service-account-name-accountid.json');
> 
> const sendEmail = async (email, subject, text) => {
>     try {
> 
>         const transporter = nodemailer.createTransport({
>             host: 'smtp.gmail.com',
>             port: 465,
>             secure: true,
>             auth: {
>                 type: 'OAuth2',
>                 user: email, //your permissioned service account member e-mail address
>                 serviceClient: json.client_id,
>                 privateKey: json.private_key
>             }
>         });
> 
>         await transporter.verify();
>         
>         await transporter.sendMail({
>                 from: json.client_email,
>                 to: email, //you can change this to any other e-mail address and it should work!
>                 subject,
>                 text
>         });
>         console.log('success!');
>         return {
>             status : 200
>         }
> 
>     } catch (error) {
>         console.log(error);
>         return {
>             status : 500,
>             error
>         }
>     }
> }
> 
> sendEmail('your_permissioned_service_account_email_address@some_place.com, 'testing 123', 'woohoo!');
> ```

I keep getting this error:

`Trace: Error: connect ETIMEDOUT 104.21.5.100:25
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1132:16) {
  errno: -4039,
  code: 'ECONNECTION',
  syscall: 'connect',
  address: '104.21.5.100',
  port: 25,
  command: 'CONN'
}`


Despite using port 465

**Comment by unigazer** [comment]
Does anyone knows why Google returns
```
Error: invalid_grant: java.security.SignatureException: Invalid signature for token: eyJhbGciOiJSUzI1NiIsInR5....
```
after some time? I have successfully set everything correctly and it was working for the first couple of days and then after some period, Google started returning the error above. I had to regenerate the service account key-pair values and the mail service started working again.

Any advice on how to prevent this from happening?

**Comment by Charly9978** [comment]
> > I got it working, these were the steps:
> > 
> > 1. Log in to console.- https://console.cloud.google.com/
> > 2. Create a service account under the project.
> > 3. Click on the new service account, go to permissions and add a member. You will use this member's email address when sending the request.
> > 4. Create keys for the service account. - keys -> add key. https://console.cloud.google.com/iam-admin/serviceaccounts
> > 5. Download your key file. You will get something like service-account-name-accountid.json. It will have all the information you need to get the code below running.
> > 6. Delegate authority to your service account https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority. Add`https://mail.google.com/` as the scope.
> > ...
> 
> I have tried your solution, but the 6th step seems to require a Google Workspace account to perform. I only have a Regular Gmail account, so this is unavailable for me.

Same issue... do you find a solution without Google workspace account ? 
Thanks

**Comment by piboistudios** [comment]
@punit1108 I got a similar issue, and resolved it by ensuring I used an actual user/principal inside of the gmail domain.

For example, the service account I am using is `no-reply@made-up-project-name.iam.gserviceaccount.com` but the `user` I am using in `auth` config object of the SMTP transport had to be `noreply@our-email-domain.com`

I guess this is because service accounts don't really have gmail accounts despite having an "email" address.

So setting up domain-wide delegation is required for this (or else your service account wont be able to authenticate as another user under `our-email-domain.com`, which again, is required, or you will get the `Invalid principal` error).

The primary caveat to this is obviously it is a security issue. If anyone gets access to that service account's credentials they can spoof emails from any user within your Google Workspace, as there doesn't appear to be a way to only authorize the service account to auth as particular users.
